### PR TITLE
write logs on disk

### PR DESCRIPTION
**Pull request by Pierregvx**
result : 
```
cat /usr/local/share/massastation/massastation.log

2023-06-28T14:37:35.500+0200    INFO    api/middleware.go:21    test
2023-06-28T14:37:35.501+0200    INFO    api/middleware.go:22    [POST /plugin-manager/register]
2023-06-28T15:28:57.400+0200    INFO    api/server.go:146       Connected to node server https://buildernet.massa.net/api/v2 (version DEVN.21.0)
2023-06-28T15:28:57.403+0200    INFO    restapi/server.go:111   Serving massastation at http://[::]:80
2023-06-28T15:28:57.404+0200    INFO    restapi/server.go:111   Serving massastation at https://[::]:443
2023-06-28T15:28:57.415+0200    INFO    api/middleware.go:21    [POST /plugin-manager/register]
2023-06-28T15:32:19.773+0200    INFO    restapi/server.go:111   Shutting down... 
2023-06-28T15:32:19.773+0200    INFO    restapi/server.go:111   Stopped serving massastation at https://[::]:443
2023-06-28T15:32:19.773+0200    INFO    restapi/server.go:111   Stopped serving massastation at http://[::]:80
2023-06-28T15:32:19.774+0200    ERROR   plugin/plugin.go:178    plugin 'station-massa-node-manager' exiting with error: signal: interrupt

2023-06-28T15:32:34.453+0200    INFO    api/server.go:146       Connected to node server https://buildernet.massa.net/api/v2 (version DEVN.21.0)
2023-06-28T15:32:34.456+0200    INFO    restapi/server.go:111   Serving massastation at http://[::]:80
2023-06-28T15:32:34.457+0200    INFO    restapi/server.go:111   Serving massastation at https://[::]:443
2023-06-28T15:32:34.462+0200    INFO    api/middleware.go:21    [POST /plugin-manager/register]
2023-06-28T15:32:40.942+0200    INFO    restapi/server.go:111   Shutting down... 
2023-06-28T15:32:40.943+0200    INFO    restapi/server.go:111   Stopped serving massastation at https://[::]:443
2023-06-28T15:32:40.943+0200    INFO    restapi/server.go:111   Stopped serving massastation at http://[::]:80
2023-06-28T15:32:40.944+0200    ERROR   plugin/plugin.go:178    plugin 'station-massa-node-manager' exiting with error: signal: interrupt
```

### CODE REVIEW

File: cmd/massastation/main.go [hunk 1]
```diff
@@ -52,7 +54,37 @@ func ParseFlags() api.StartServerFlags {
...
+
+	wrt := io.MultiWriter(os.Stdout, f)
+	log.SetOutput(wrt)
+	return &logger{f, wrt}
+}
+
+func (l *logger) close() {
+	l.f.Close()
+	l.wrt = nil
+}
```
Comment (gregLibert) [inline cmd/massastation/main.go:83]:
```suggestion
func newLogger() *logger {
	logDir, err := config.GetConfigDir()
	if err != nil {
		log.Fatal(err)
	}

	logFilePath := filepath.Join(logDir, "massastation.log")
	
	rawJSON := []byte(`{
	  "level": "info",
	  "encoding": "json",
	  "outputPaths": ["stdout", "` + logFilePath + `"],
	  "errorOutputPaths": ["stderr", "` + logFilePath + `"]
	}`)

	var cfg zap.Config
	if err := json.Unmarshal(rawJSON, &cfg); err != nil {
		panic(err)
	}
	return zap.Must(cfg.Build())
}
```

File: cmd/massastation/main.go [hunk 2]
```diff
@@ -52,7 +54,37 @@ func ParseFlags() api.StartServerFlags {
...
+
 func main() {
+	logger := newLogger()
+	defer logger.close()
```
Comment (gregLibert) [inline cmd/massastation/main.go:87]:
```suggestion
	defer logger.Sync()
```

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: cmd/massastation/main.go [hunk 3]
```diff
@@ -21,18 +21,15 @@ func ParseFlags() api.StartServerFlags {
...
-			`MassaStation can't run without a config directory.
-			Please reinstall MassaStation using the installer at https://github.com/massalabs/station and try again.`,
-		)
+		config.Logger.Error(fmt.Sprintf("Unable to read config dir: %s\n%s", err, "MassaStation can't run without a config directory.\nPlease reinstall MassaStation using the installer at https://github.com/massalabs/station and try again."))
 	}
 
 	certDir, err := config.GetCertDir()
 	if err != nil {
-		log.Fatalln("Unable to read cert dir:", err,
+		config.Logger.Error(fmt.Sprintf("Unable to read cert dir:%s\n%s", err,
```
Comment (peterjah) [inline cmd/massastation/main.go:29]: i think this is not iso functionnal. Fatalln will exit the app. and i dont think config.Logger.Error wil do the same.
or you keep log.Fatalln here. or you have to add a  os.Exit(1) after this line

File: cmd/massastation/main.go [hunk 4]
```diff
@@ -53,15 +50,17 @@ func ParseFlags() api.StartServerFlags {
...
 	if flags.Version {
-		log.Println("Version:", config.Version)
+		config.Logger.Info(fmt.Sprintf("Version:%s", config.Version))
 		os.Exit(0)
 	}
 
 	networkManager, err := config.NewNetworkManager()
 	if err != nil {
-		log.Fatal("Failed to create NetworkManager:", err)
+		config.Logger.Error(fmt.Sprintf("Failed to create NetworkManager:%s", err))
```
Comment (peterjah) [inline cmd/massastation/main.go:63]: same here

**Review by peterjah — CHANGES_REQUESTED** [review]
little thing to fix, but good job!

**Review by peterjah — APPROVED** [review]

File: cmd/massastation/main.go [hunk 5]
```diff
@@ -21,18 +21,15 @@ func ParseFlags() api.StartServerFlags {
 
 	_, err := config.GetConfigDir()
 	if err != nil {
-		log.Fatalln("Unable to read config dir:", err,
-			`MassaStation can't run without a config directory.
-			Please reinstall MassaStation using the installer at https://github.com/massalabs/station and try again.`,
-		)
+		config.Logger.Error(fmt.Sprintf("Unable to read config dir: %s\n%s", err, "MassaStation can't run without a config directory.\nPlease reinstall MassaStation using the installer at https://github.com/massalabs/station and try again."))
```
Comment (gregLibert) [inline cmd/massastation/main.go:24]: Errorf ?

File: cmd/massastation/main.go [hunk 6]
```diff
@@ -21,18 +21,15 @@ func ParseFlags() api.StartServerFlags {
...
 	}
 
 	certDir, err := config.GetCertDir()
 	if err != nil {
-		log.Fatalln("Unable to read cert dir:", err,
+		config.Logger.Fatal(fmt.Sprintf("Unable to read cert dir:%s\n%s", err,
```
Comment (gregLibert) [inline cmd/massastation/main.go:29]: Fatalf ?

File: int/api/server.go [hunk 7]
```diff
@@ -83,7 +83,7 @@ type Server struct {
 func NewServer(flags StartServerFlags) *Server {
 	swaggerSpec, err := loads.Analyzed(restapi.SwaggerJSON, "")
 	if err != nil {
-		log.Fatalln(err)
+		config.Logger.Error(err.Error())
```
Comment (gregLibert) [inline int/api/server.go:86]: should be a fatal

File: int/api/server.go [hunk 8]
```diff
@@ -113,22 +113,22 @@ func (server *Server) Start(networkManager *config.NetworkManager) {
...
 
-	log.Println("Server started")
+	config.Logger.Debug("Server started")
 }
 
 // Stops the server and waits for it to finish.
 func (server *Server) Stop() {
 	if err := server.api.Shutdown(); err != nil {
-		log.Fatalln(err)
+		config.Logger.Error(err.Error())
```
Comment (gregLibert) [inline int/api/server.go:126]: should be a fatal

File: pkg/config/network.go [hunk 9]
```diff
@@ -146,19 +145,19 @@ func (n *NetworkManager) Network() *AppConfig {
...
-		DNSAddress: config.DNS,
+		NodeURL:    cfg.URLs[0],
+		DNSAddress: cfg.DNS,
 		Network:    selectedNetworkStr,
 	})
 
-	log.Printf("Switched to network: %s\n", selectedNetworkStr)
-	log.Printf("Current config: %+v\n", n.Network())
+	Logger.Info(fmt.Sprintf("Switched to network: %s\n", selectedNetworkStr))
+	Logger.Info(fmt.Sprintf("Current config: %+v\n", n.Network()))
```
Comment (gregLibert) [inline pkg/config/network.go:160]: Should be debug

File: pkg/convert/byteConverter.go [hunk 10]
```diff
@@ -92,7 +94,7 @@ func BytesToU64(byteArray []byte) uint64 {
 
 	err := binary.Read(bytes.NewReader(byteArray), binary.LittleEndian, &u64)
 	if err != nil {
-		log.Printf("error converting bytesToU64 :%v\n", err)
+		config.Logger.Info(fmt.Sprintf("error converting bytesToU64 :%v\n", err))
```
Comment (gregLibert) [inline pkg/convert/byteConverter.go:97]: should be an error

File: pkg/plugin/plugin.go [hunk 11]
```diff
@@ -175,14 +175,14 @@ func (p *Plugin) Start() error {
...
-			log.Printf("plugin '%s' exiting with error: %s\n", pluginName, err)
+			config.Logger.Error(fmt.Sprintf("plugin '%s' exiting with error: %s\n", pluginName, err))
 
 			p.status = Crashed
 
 			return
 		}
 
-		log.Printf("plugin '%s' exiting without error.\n", pluginName)
+		config.Logger.Info(fmt.Sprintf("plugin '%s' exiting without error.\n", pluginName))
```
Comment (gregLibert) [inline pkg/plugin/plugin.go:185]: should be a debug

File: pkg/plugin/plugin.go [hunk 12]
```diff
@@ -192,7 +192,7 @@ func (p *Plugin) Start() error {
 
 // Kills a plugin.
 func (p *Plugin) Stop() error {
-	log.Printf("Stopping plugin %s.\n", p.ID)
+	config.Logger.Info(fmt.Sprintf("Stopping plugin %s.\n", p.ID))
```
Comment (gregLibert) [inline pkg/plugin/plugin.go:195]: should be a debug

File: pkg/store/store.go [hunk 13]
```diff
@@ -132,10 +131,10 @@ func (s *Store) FetchStorePeriodically() {
 	for range ticker.C {
 		err := s.FetchPluginList()
 		if err != nil {
-			log.Printf("while fetching plugin list: %s", err)
+			config.Logger.Error(fmt.Sprintf("while fetching plugin list: %s", err))
 		}
 
-		log.Printf("Fetched plugin list. %d plugins in store.", len(s.Plugins))
+		config.Logger.Info(fmt.Sprintf("Fetched plugin list. %d plugins in store.", len(s.Plugins)))
```
Comment (gregLibert) [inline pkg/store/store.go:137]: should be a debug

**Review by gregLibert — CHANGES_REQUESTED** [review]
You should replace fmt;Sprintf if possible by uisng XXXf functions. the code will be simpler to read.

**Comment by Pierregvx** [comment]
> You should replace fmt;Sprintf if possible by uisng XXXf functions. the code will be simpler to read.

I couldn't see these functions in https://github.com/uber-go/zap/blob/master/logger.go
but maybe I can use an addition ?

**Review by gregLibert — APPROVED** [review]

File: pkg/config/logger.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+	"path/filepath"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+//nolint:gochecknoglobals
+var Logger *zap.Logger
+
+func NewLogger() *zap.Logger {
```
[the remark this answers is not in the document]
Comment (gregLibert) [reply pkg/config/logger.go]: ok, I cleaned it to much, my bad...

File: pkg/config/logger.go [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+//nolint:gochecknoglobals
+var Logger *zap.SugaredLogger
+
+func NewLogger() *zap.SugaredLogger {
+	logDir, err := GetConfigDir()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	logFilePath := filepath.Join(logDir, "massastation.log")
```
Comment (Thykof) [inline pkg/config/logger.go:21]: logFilePath := filepath.Join(logDir, "logs, "massastation.log")

**Pull request merged**